manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,443 @@
+in OCaml where the type inferred by the type system may be less generic
+than expected. Such non-genericity can stem from interactions between
+side-effect and typing or the difficulties of implicit polymorphic
+recursion or higher-rank polymorphism.
+
+This chapter details each of these situations and, if it is possible,
+how to recover genericity.
+
+\section{Weak polymorphism and mutation}
+\subsection{Weakly polymorphic types}
+\label{ss:weaktypes}
+Maybe the most frequent examples of non-genericity derive from the
+interactions between polymorphic types and mutation. A simple example
+appears when typing the following functions
+
+\begin{caml_example}
+let store = ref None ;;
+\end{caml_example}
+
+Since the type of "None" is "'a option" and the function "ref" has type
+"'b -> 'b ref", a natural deduction for the type of "store" would be
+"'a option ref". However, the inferred type, "'_a option ref", is slightly
+different. Type variables whose name starts with an "_" like "'_a" are weakly
+polymorphic type variables, sometimes shortened as weak type variables. A weak
+type variable is a placeholder for a single type that is currently unknown.
+Once the specific type "t" behind the placeholder type "'_a" known, all
+occurrences of "'_a" will be replaced by "t".
+
+This distinction between weakly and generic polymorphic type variable protects
+OCaml program from unsoundness and runtime errors. To understand from where
+unsoundness might come, consider this simple function which swaps a value "x"
+with the value stored inside a "store" reference, if there is such value:
+\begin{caml_example}
+let swap store x = match !store with
+  | None -> store := Some x; x
+  | Some y -> store := Some x; y;;
+\end{caml_example}
+We can then apply this function to our store
+\begin{caml_example}
+let one = swap store 1
+let one_again = swap store 2
+let two = swap store 3;;
+\end{caml_example}
+After these three swaps the stored value is "3". Everything is fine up to
+now. We can then try to swap "2" with a more interesting value, for
+instance a function:
+\begin{caml_example}
+let error = swap store (fun x -> x);;
+\end{caml_example}
+At this point, the type checker rightfully complains that it is not
+possible to swap an integer and a function, and that an "int" should always
+be traded for another "int". Furthermore, the type checker prevents us to
+change manually the value stored by "store":
+\begin{caml_example}
+store := Some (fun x -> x);;
+\end{caml_example}
+Indeed, looking at the type of store, we see that the weak type "'_a" has
+been replaced by the type "int"
+\begin{caml_example}
+store;;
+\end{caml_example}
+Therefore, after placing an "int" in "store", we cannot use it to store any
+value other than an "int". More generally, weak types protect the program from
+undue mutation of values with a polymorphic type.
+
+%todo: fix indentation in pdfmanual
+Moreover, weak type cannot appear as the type of toplevel values:
```

Comment:
I would rather say that weak types cannot appear in signatures in order
to include cases like
```
module M = struct module N = struct let store = ref None end end
```